tests/federation_room_join_partial_state_test.go

```diff
@@ -182,6 +182,104 @@ func TestPartialStateJoin(t *testing.T) {
+		// set up 3 homeservers: hs1, hs2 and complement
+		deployment := Deploy(t, b.BlueprintFederationTwoLocalOneRemote)
+		defer deployment.Destroy(t)
+		alice := deployment.Client(t, "hs1", "@alice:hs1")
+		charlie := deployment.Client(t, "hs2", "@charlie:hs2")
+
+		// create a public room
+		roomID := alice.CreateRoom(t, map[string]interface{}{
+			"preset": "public_chat",
+		})
+
+		// create the complement homeserver
+		server := federation.NewServer(t, deployment,
+			federation.HandleKeyRequests(),
+			federation.HandlePartialStateMakeSendJoinRequests(),
+			federation.HandleEventRequests(),
+			federation.HandleTransactionRequests(
+				func(e *gomatrixserverlib.Event) {
+					t.Fatalf("Received unexpected PDU: %s", string(e.JSON()))
+				},
+				// hs1 may send us presence when alice syncs
+				func(e gomatrixserverlib.EDU) {},
+			),
+		)
+		cancelListener := server.Listen()
+		defer cancelListener()
+
+		// join complement to the public room
+		server.MustJoinRoom(t, deployment, "hs1", roomID, server.UserID("bob"))
+
+		// we expect a /state_ids request from hs2 after it joins the room
+		// we will respond to the request with garbage
+		requestReceivedWaiter := NewWaiter()
+		sendResponseWaiter := NewWaiter()
+		server.Mux().Handle(
+			fmt.Sprintf("/_matrix/federation/v1/state_ids/%s", roomID),
+			http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+				queryParams := req.URL.Query()
+				t.Logf("Incoming state_ids request for event %s in room %s", queryParams["event_id"], roomID)
+				requestReceivedWaiter.Finish()
+				sendResponseWaiter.Waitf(t, 60*time.Second, "Waiting for /state request")
+				t.Logf("Replying to /state_ids request with invalid response")
+
+				w.WriteHeader(200)
+
+				if _, err := w.Write([]byte("{}")); err != nil {
+					t.Errorf("Error writing to request: %v", err)
+				}
+			}),
+		).Methods("GET")
+
+		// join charlie on hs2 to the room, via the complement homeserver
+		charlie.JoinRoom(t, roomID, []string{server.ServerName()})
+
+		// and let hs1 know that charlie has joined,
+		// otherwise hs1 will refuse /state_ids requests
+		room := server.Room(roomID)
+		member_event := room.CurrentState("m.room.member", charlie.UserID).JSON()
+		server.MustSendTransaction(t, deployment, "hs1", []json.RawMessage{member_event}, nil)
+		alice.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(charlie.UserID, roomID))
+
+		// wait until hs2 starts syncing state
+		requestReceivedWaiter.Waitf(t, 5*time.Second, "Waiting for /state_ids request")
+
+		syncResponseChan := make(chan gjson.Result)
+		defer close(syncResponseChan)
+		go func() {
+			response, _ := charlie.MustSync(t, client.SyncReq{})
+			syncResponseChan <- response
+		}()
+
+		// the client-side requests should still be waiting
+		select {
+		case <-syncResponseChan:
+			t.Fatalf("hs2 sync completed before state resync complete")
+		default:
+		}
+
+		// reply to hs2 with a bogus /state_ids response
+		sendResponseWaiter.Finish()
```

Comment:
This could have a clearer name, e.g `sendStateIDsResponse`, as otherwise it's unclear which response you mean and need to add comments to explain it.

Reply:
I'll rename them to follow the naming in `partialStateJoinResult`:
```go
type partialStateJoinResult struct {
	cancelListener                   func()
	Server                           *federation.Server
	ServerRoom                       *federation.ServerRoom
	fedStateIdsRequestReceivedWaiter *Waiter
	fedStateIdsSendResponseWaiter    *Waiter
}
```